src/ast/ast_validations.c

```diff
@@ -1509,6 +1510,70 @@ static AST_Validation _BlockUnsupportedMerges(AST *ast, char **reason) {
 	return res;
 }
 
+static void _colect_query_parameters_names(const cypher_astnode_t *root, rax *keys) {
+	cypher_astnode_type_t type = cypher_astnode_type(root);
+	// In case of identifier.
+	if(type == CYPHER_AST_PARAMETER) {
+		const char *identifier = cypher_ast_parameter_get_name(root);
+		raxInsert(keys, (unsigned char *)identifier, strlen(identifier), NULL, NULL);
+	} else {
+		// Recurse over children.
+		uint child_count = cypher_astnode_nchildren(root);
+		for(uint i = 0; i < child_count; i++) {
+			const cypher_astnode_t *child = cypher_astnode_get_child(root, i);
+			// Recursively continue mapping.
+			_colect_query_parameters_names(child, keys);
+		}
+	}
+}
+
+// This method extracts given parameters names.
+static rax *_extract_given_params(const cypher_astnode_t *statement) {
+	rax *given_params_names = raxNew();
+	uint noptions =  cypher_ast_statement_noptions(statement);
+	for(uint i = 0; i < noptions; i++) {
+		const cypher_astnode_t *option = cypher_ast_statement_get_option(statement, i);
+		uint nparams = cypher_ast_cypher_option_nparams(option);
+		for(uint j = 0; j < nparams; j++) {
+			const cypher_astnode_t *param = cypher_ast_cypher_option_get_param(option, j);
+			const char *paramName = cypher_ast_string_get_value(cypher_ast_cypher_option_param_get_name(param));
+			raxInsert(given_params_names, (unsigned char *) paramName, strlen(paramName), NULL, NULL);
```

Comment:
By checking the return value here, we can add a validation error on shadowed parameters:
`CYPHER intval=30 intval=31 RETURN $intval`

Reply:
done